autoneg.go

```diff
@@ -76,63 +73,72 @@ func (accept accept_slice) Less(i, j int) bool {
 	return false
 }
 
-func (accept accept_slice) Swap(i, j int) {
-	slice := []Accept(accept)
+func (slice acceptSlice) Swap(i, j int) {
 	slice[i], slice[j] = slice[j], slice[i]
 }
 
+// trimSpaces trims spaces from the given string.
+func trimSpaces(s string) string {
+	if len(s) == 0 || (s[0] != ' ' && s[len(s)-1] != ' ') {
+		// strings.Trim performs memory allocations, so implement
+		// fast path to avoid those if not needed.
+		return s
+	}
+	return strings.Trim(s, " ")
+}
+
 // Parse an Accept Header string returning a sorted list
 // of clauses
-func ParseAccept(header string) (accept []Accept) {
+func ParseAccept(header string) []Accept {
 	parts := strings.Split(header, ",")
-	accept = make([]Accept, 0, len(parts))
-	for _, part := range parts {
-		part := strings.Trim(part, " ")
-
-		a := Accept{}
-		a.Params = make(map[string]string)
-		a.Q = 1.0
-
-		mrp := strings.Split(part, ";")
+	accept := make(acceptSlice, 0, len(parts))
```

Comment:
You can add another fast path here by not splitting (which allocates) if there's no comma.

Reply:
It would actually be better to make this a loop that tokenizes parts on `,` if you want to avoid allocations.  I did that in go-restful to bypass https://github.com/emicklei/go-restful/blob/95fec025ab35f1d26407d1a4ef1542a3bf4eb8a0/route.go#L79

Reply:
I was optimizing the k8s usage TBH. And for all calls from k8s client (vast majority) there is a comma actually:
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/rest/request.go#L140

> It would actually be better to make this a loop that tokenizes parts on , if you want to avoid allocations. I did that in go-restful to bypass https://github.com/emicklei/go-restful/blob/95fec025ab35f1d26407d1a4ef1542a3bf4eb8a0/route.go#L79

I did part of it for part, but it probably makes sense to go further. Will change that.